test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java

```diff
@@ -0,0 +1,145 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio.channel;
+
+import org.elasticsearch.transport.nio.ESSelector;
+import org.elasticsearch.transport.nio.SocketSelector;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.SocketChannel;
+
+public class NioSocketChannel extends AbstractNioChannel<SocketChannel> {
+
+    private final InetSocketAddress remoteAddress;
+    private final ConnectFuture connectFuture = new ConnectFuture();
+    private WriteContext writeContext;
+    private ReadContext readContext;
+    private SocketSelector socketSelector;
+
+    public NioSocketChannel(String profile, SocketChannel socketChannel) throws IOException {
+        super(profile, socketChannel);
+        this.remoteAddress = (InetSocketAddress) socketChannel.getRemoteAddress();
+    }
+
+    @Override
+    public CloseFuture close(boolean attemptToCloseImmediately) {
+        // Even if the channel has already been closed we will clear any pending write operations just in case
+        if (attemptToCloseImmediately && state.get() > UNREGISTERED && selector.isOnCurrentThread() && writeContext.hasQueuedWriteOps()) {
+            writeContext.clearQueuedWriteOps(new ClosedChannelException());
+        }
+
+        return super.close(attemptToCloseImmediately);
+    }
+
+    @Override
+    public SocketSelector getSelector() {
+        return socketSelector;
+    }
+
+    @Override
+    public boolean markRegistered(ESSelector selector) {
+        this.socketSelector = (SocketSelector) selector;
+        return super.markRegistered(selector);
+    }
+
+    public int write(ByteBuffer buffer) throws IOException {
+        return socketChannel.write(buffer);
+    }
+
+    public long vectorizedWrite(ByteBuffer[] buffers) throws IOException {
+        return socketChannel.write(buffers);
+    }
+
+    public int read(ByteBuffer buffer) throws IOException {
+        return socketChannel.read(buffer);
+    }
+    public WriteContext getWriteContext() {
+        return writeContext;
+    }
+
+    public void setReadContext(TcpReadContext readContext) {
+        this.readContext = readContext;
+    }
+
+    public ReadContext getReadContext() {
+        return readContext;
+    }
+
+    public void setWriteContext(TcpWriteContext writeContext) {
+        this.writeContext = writeContext;
+    }
+
+    public InetSocketAddress getRemoteAddress() {
+        return remoteAddress;
+    }
+
+    public boolean isConnectComplete() {
+        return connectFuture.isConnectComplete();
+    }
+
+    public boolean isWritable() {
+        return state.get() == REGISTERED;
+    }
+
+    public boolean isReadable() {
+        return state.get() == REGISTERED;
+    }
+
+    public boolean finishConnect() throws IOException {
```

Comment:
I'm not sure. `finishConnect()` a method that you have to call on the real SocketChannel to ensure that it is ready to read and write. In this case, this method is making that call + doing other things necessary to ensure that our channel abstraction is set up (and futures are completed). 

And we do not really "wait" at any point. If the raw "finishConnect()" method returns false well wait until a OP_CONNECT event to try again.